addons/mail/static/src/core/common/message_actions.js

```diff
@@ -196,7 +197,19 @@ messageActionsRegistry
         icon: (component) =>
             `fa fa-language ${component.state.showTranslation ? "o-mail-Message-translated" : ""}`,
         title: (component) => (component.state.showTranslation ? _t("Revert") : _t("Translate")),
-        onClick: (component) => component.onClickToggleTranslation(),
+        onClick: async (component) => {
+            const message = toRaw(component.message);
+            if (!message.translationValue) {
+                const { error, lang_name, body } = await rpc("/mail/message/translate", {
+                    message_id: message.id,
+                });
+                message.translationValue = body && markup(body);
+                message.translationSource = lang_name;
+                message.translationErrors = error;
+            }
+            component.state.showTranslation =
+                !component.state.showTranslation && Boolean(message.translationValue);
```

Comment:
I'm not super happy that this duplicates code between mobile and desktop.
@alexkuhn we should move the translation logic inside the message model. That way, we can centralize the translation code and avoid forcing components to implement methods.